Refactor the user creation / user profile form

roshnaeem opened the pull request:
**Description:**
1. Created three new files to provide the same functionality `user.pl` and `user-form.tt.html` were providing.
2. `user.pl` uses three functions defined in `Users.pm` for user creation and user profile form. These three functions consist highly of HTML.
3. The newly created three files which can be viewed in this PR are using Template::Toolkit to provide the same functionality and remove the usage of HTML including functions. 
4. Moreover, removed the usage of `<table>` to make the form responsive.



**Related issues and discussion:** https://github.com/openfoodfacts/openfoodfacts-server/issues/5106

stephanegigandet commented on templates/user_form2.tt.html:
```diff
@@ -0,0 +1,241 @@
+    <form method="post" action="/cgi/user2.pl" enctype="multipart/form-data">
+		[% FOREACH section IN sections %]
+			[% IF user OR section.id == "user" %]
+				[% FOREACH field IN section.fields %]
```
Instead of having 3 different [% FOREACH field IN section.fields %] loops, for each of the sections, it would be better to keep only one that works for all 3 sections, like it is done for the org.pl form:

		[% FOREACH section IN sections %]
		
			[% IF admin OR section.id != "admin" %]
		
				[% IF section.name %]
					<fieldset[% IF section.id == "admin" %] class="panel callout"[% END %]>
						<legend>[% section.name %]</legend>
				[% END %]
				
				[% IF section.description %]
					<p>[% section.description %]</p>
				[% END %]
				[% IF section.note %]
					<p>[% section.note %]</p>
				[% END %]

				[% FOREACH field IN section.fields %]

					[% IF field.description %]
						<p>[% field.description %]</p>
					[% END %]								

					[% IF field.type == 'text' %]
						<label for="[% field.field %]">[% field.label %]</label>
						<input type="text" id="[% field.field %]" name="[% field.field %]" value="[% field.value %]" />
					[% ELSIF field.type == 'textarea' %]
						<label for="[% field.field %]">[% field.label %]</label>
						<textarea id="[% field.field %]" name="[% field.field %]" style="height:100px;">[% field.value %]</textarea>
					[% ELSIF field.type == 'checkbox' %]
						<label for="[% field.field %]">
							<input type="checkbox" id="[% field.field %]" name="[% field.field %]" [% IF field.value == 'on' %]checked="checked"[% END %] />
							[% field.label %]
						</label>
					[% END %]
					
					[% IF field.note %]
						<p style="font-size:.8em">&rarr; [% field.note %]</p>
					[% END %]

				[% END %]
				
				[% IF section.name %]
					</fieldset>
					[% IF section.name == "admin" %]
						</div>
					[% END %]
				[% END %]
			
			[% END %]

		[% END %]

roshnaeem replied:
Actually in `user.pl` the msgctxt used from `.po` files have different names, they are not following a rule like in `org.pl`. `org.pl` have msgctxt ending with `_note` or `_description`. But they are different for each section in `user.pl`, so we can not add them in the start here for every section. That's why I have used `foreach` separately for each section so things don't get mixed up.

stephanegigandet replied:
For the fields where the "_note" or "_description" does not work, in user2.pl, you could still assign the value in the user2.pl file:

				{
					field => "requested_org",
					label => lang("producer_or_brand"),
				}

stephanegigandet commented on templates/user_form2.tt.html:
```diff
@@ -0,0 +1,241 @@
+
+    <form method="post" action="/cgi/user2.pl" enctype="multipart/form-data">
+		[% FOREACH section IN sections %]
+			[% IF user OR section.id == "user" %]
```
[% IF user OR section.id == "user" %]

--> there's no "user" variable in the template data $template_ref

in the org template, there is a similar line, but only for administrators:

[% IF admin OR section.id != "admin" %]

--> this is just to hide the "admin" section for administrators.  The "admin" variable is set in the template_data.

roshnaeem replied:
`user` is the section id for the first section. every section has a different unique ID. These IDs are helping to distinguish the sections from each other.

stephanegigandet commented on cgi/user2.pl:
```diff
@@ -0,0 +1,361 @@
+		for (my $i = 1; $i <= 3; $i++) {
+			push @{$team_section_ref->{fields}}, {
+				field => "team_". $i,
+				label2 => sprintf(lang("team_s"), $i),
```
Why not use the "label" instead of a new "label2" ?

roshnaeem replied:
We are setting the `label` equal to `lang($field_lang_id)` if it is defined. I checked setting it to `label` instead of `label2`, but it applies `lang` function twice on it resulting in bugs.

stephanegigandet commented on cgi/user2.pl:
```diff
@@ -0,0 +1,361 @@
+		foreach my $group (@user_groups) {
+			push @{$administrator_section_ref->{fields}}, {
+				field =>  "user_group_". $group,
+				label2 =>  "user_group_" . ${group} . "_description",
```
in the org template, there is a "description" that could be used for what you put in "label2"

e.g. in org.pl it's populated automatically if we have a "[fieldid]_description" entry in the .po files:

			# Label
			$field_ref->{label} = lang($field_lang_id);
			
			# Descriptions and notes for fields
			if (lang($field_lang_id . "_description")) {
				$field_ref->{description} = lang($field_lang_id . "_description");
			}
			if (lang($field_lang_id . "_note")) {
				$field_ref->{note} = lang($field_lang_id . "_note");
			}

roshnaeem replied:
yes you are right, we can use the description here instead of `label2` in the administrator section.

stephanegigandet commented on templates/user_form2.tt.html:
```diff
@@ -0,0 +1,241 @@
+                [% ELSIF product_type.defined && product_type == "food" %]
+                    [% FOREACH field IN section.fields %]
+                        [% IF field.type == 'checkbox' %]
+                            <input type="checkbox" id="[% field.field %]" name="[% field.field %]" [% IF pro_checked == 'on' %]checked="checked"[% END %] />
```
Here you could do it like in the org template:

						<label for="[% field.field %]">
							<input type="checkbox" id="[% field.field %]" name="[% field.field %]" [% IF field.value == 'on' %]checked="checked"[% END %] />
							[% field.label %]
						</label>

roshnaeem replied:
The issue here was that in `users.pm` it is used like this. The label is associated with a hidden field. I first tried it like this, but it was causing issues of different names I guess.  So I tried to stick to the `Users.pm` to avoid errors.

Here is the code snippet of `Users.pm`:

`"<input type=\"checkbox\" id=\"pro\" name=\"pro\" $pro_checked>"
		. "<label for=\"pro\">"
		. "<input type=\"hidden\" name=\"pro_checkbox\" value=\"1\"> "
		. lang("this_is_a_pro_account") . "</label>"`

stephanegigandet commented on cgi/user2.pl:
```diff
@@ -0,0 +1,361 @@
+	$template_data_ref->{requested_org_ref} = $requested_org_ref;
+	$template_data_ref-> {org_name} = sprintf(lang("add_user_existing_org"), org_name($requested_org_ref));
+	$template_data_ref->{teams_flag} = not ((defined $server_options{private_products}) and ($server_options{private_products}));
+	$template_data_ref->{admin_flag} = $admin;
```
There is already a "admin" variable that is automatically set by process_template():

sub process_template($$$) {
	
	my $template_filename = shift;
	my $template_data_ref = shift;
	my $result_content_ref = shift;
	
	# Add functions and values that are passed to all templates

	$template_data_ref->{product_type} = $options{product_type};
	$template_data_ref->{admin} = $admin;

roshnaeem replied:
Alright, `$admin` was causing me a lot of issues, and I did not where it was being set. That is why I changed it to `$admin_flag`. Let me check it again how can I fix it.

stephanegigandet commented on cgi/user2.pl:
```diff
@@ -0,0 +1,361 @@
+		$pro_checked = "on";
+	}
+
+	$template_data_ref->{product_type} = $options{product_type};
```
this is already set by process_template()

roshnaeem replied:
Let me check this as well. Sorry, I did not explore `process_template()` which I should have.

stephanegigandet commented:
Hello @roshnaeem !

Thanks a lot for the PR. I tested it and it is working as expected in all the cases that I could think of.

I think the template code can be made much more generic though, with less hardcoded values and special cases for each section.

Here there is a lot of logic that is now in the template. We should keep the logic in the Perl code, and put the result of the logic in the template_data (the different sections and the different fields for each section). Then the template can decide how to display them (e.g. whether to use a table, or fieldsets etc.).

That way, if we had a new section, we should not have to modify the template.

Also we should try to make the template as close as possible to the template used for the org.pl form. At some point we could even see if we could have one "form" template that we could use for all similar forms.

roshnaeem commented on templates/user_form2.tt.html:
```diff
@@ -0,0 +1,241 @@
+                        <input type="text" id="[% field.field %]" name="[% field.field %]" value="[% field.value %]" autocomplete="[% field.field %]"/>
+					[% ELSIF field.type == 'password' %]
+						<label for="[% field.field %]">[% lang(field.label2) %]</label>
+						<input type="password" id="[% field.field %]" name="[% field.field %]" value="[% user_password_field %]" autocomplete="new-password"/>
```
` "\n<tr><td>$Lang{password}{$lang}</td><td>"
	. password_field(-name=>'password', -value=>$user_ref->{password}, -autocomplete=>'new-password', -override=>1) . "</td></tr>"
	. "\n<tr><td>$Lang{password_confirm}{$lang}</td><td>"
	. password_field(-name=>'confirm_password', -value=>$user_ref->{password}, -autocomplete=>'new-password', -override=>1) . "</td></tr>"`

This is the code snippet of `Users.pm`, here for both `password` and `confirm_password` value is same. That's why I stored it in one variable and used it for both fields. 

But I guess it can be replaced.

roshnaeem commented:
> Hello @roshnaeem !
> 
> Thanks a lot for the PR. I tested it and it is working as expected in all the cases that I could think of.
> 
> I think the template code can be made much more generic though, with less hardcoded values and special cases for each section.
> 
> Here there is a lot of logic that is now in the template. We should keep the logic in the Perl code, and put the result of the logic in the template_data (the different sections and the different fields for each section). Then the template can decide how to display them (e.g. whether to use a table, or fieldsets etc.).
> 
> That way, if we had a new section, we should not have to modify the template.
> 
> Also we should try to make the template as close as possible to the template used for the org.pl form. At some point we could even see if we could have one "form" template that we could use for all similar forms.

Hello Stephane, I have updated the files and incorporated all changes, Please have a look.

stephanegigandet reviewed (approved):
Thanks Rosheen!

It is a huge improvement to what we had before, it's great that the form is now responsive and that a lot of HTML code has been moved to the template.

There are a few things that I think could be improved, but it's something that we can do later, or that maybe I can do on top of your PR.

So let's merge this, deploy it and try it out (as you made a separate user2.pl script it will be easy to test / compare with what we had before).

Thank you!
